### PR TITLE
Define FileMap; fix #35

### LINKED ISSUE #35: What are "filemaps"?
They're mentioned on <https://rust-lang-nursery.github.io/rustc-guide/the-parser.html>, but not defined. I think it would be good to explain what they are first.

### PR DESCRIPTION
**Pull request by mark-i-m**
cc #48 @alexreg

### CODE REVIEW

**Comment by mark-i-m** [comment]
@alexreg If this looks good to you I can go ahead and merge it :)

File: src/appendix/code-index.md [hunk 1]
```diff
@@ -9,12 +9,13 @@ Item            |  Kind    | Short description           | Chapter            |
...
 `DefId` | struct | One of four types of HIR node identifiers. | [Identifiers in the HIR] | [src/librustc/hir/def_id.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/def_id/struct.DefId.html)
 `DiagnosticBuilder` | struct | A struct for building up compiler diagnostics, such as errors or lints | [Emitting Diagnostics] | [src/librustc_errors/diagnostic_builder.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.DiagnosticBuilder.html)
 `DocContext` | struct | A state container used by rustdoc when crawling through a crate to gather its documentation | [Rustdoc] | [src/librustdoc/core.rs](https://github.com/rust-lang/rust/blob/master/src/librustdoc/core.rs)
-`ast::Crate` | struct | Syntax-level representation of a parsed crate | [The parser] | [src/librustc/hir/mod.rs](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/struct.Crate.html)
+`FileMap` | struct | A single source within a `CodeMap` (e.g. the source code within a single file). | [The parser] | [src/libsyntax_pos/lib.rs](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/codemap/struct.FileMap.html)
```
Comment (alexreg) [inline src/appendix/code-index.md]: `A single file` sounds better, maybe?

File: src/appendix/code-index.md [hunk 2]
```diff
@@ -9,12 +9,13 @@ Item            |  Kind    | Short description           | Chapter            |
...
+`hir::Crate` | struct | More abstract, compiler-friendly form of a crate's AST | [The Hir] | [src/librustc/hir/mod.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/struct.Crate.html)
 `DefId` | struct | One of four types of HIR node identifiers. | [Identifiers in the HIR] | [src/librustc/hir/def_id.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/def_id/struct.DefId.html)
 `DiagnosticBuilder` | struct | A struct for building up compiler diagnostics, such as errors or lints | [Emitting Diagnostics] | [src/librustc_errors/diagnostic_builder.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.DiagnosticBuilder.html)
 `DocContext` | struct | A state container used by rustdoc when crawling through a crate to gather its documentation | [Rustdoc] | [src/librustdoc/core.rs](https://github.com/rust-lang/rust/blob/master/src/librustdoc/core.rs)
-`ast::Crate` | struct | Syntax-level representation of a parsed crate | [The parser] | [src/librustc/hir/mod.rs](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/struct.Crate.html)
```
Comment (alexreg) [inline src/appendix/code-index.md]: Real nitpick: start with "A syntax-level representation", for consistency? There may be other places this isn't followed either.

File: src/appendix/code-index.md (hunk 1 above)
Comment (mark-i-m) [reply src/appendix/code-index.md]: You mean instead of "A single source"?

Comment (alexreg) [reply src/appendix/code-index.md]: Yes, I suppose... I'm still a little confused though. Does it map AST nodes to source code within a particular file?

File: src/appendix/code-index.md (hunk 2 above)
Comment (alexreg) [reply src/appendix/code-index.md]: Another one: some bullet points end in full stops, some don't.

File: src/appendix/code-index.md (hunk 1 above)
Comment (mark-i-m) [reply src/appendix/code-index.md]: Better now?

Comment (alexreg) [reply src/appendix/code-index.md]: Okay. Maybe write a description parallel to that for `CodeMap` then? e.g. "A `FileMap` maps the AST nodes to their source code within a single file (i.e. like a `CodeMap`, but more restricted to one file)". Then you can changed the description for `CodeMap` to something like "A `CodeMap` maps the AST nodes to their source code (i.e. like a `FileMap`, but for an entire crate)"... I think those definitions are correct?

File: src/appendix/code-index.md (hunk 2 above)
Comment (mark-i-m) [reply src/appendix/code-index.md]: > Real nitpick: start with "A syntax-level representation", for consistency? There may be other places this isn't followed either.

Done.

> Another one: some bullet points end in full stops, some don't.

Yes, the annoying part is that not all of the summaries are complete sentences. I removed the full stops when they occur at the end.

Comment (alexreg) [reply src/appendix/code-index.md]: Yeah, true. Ideally they should have consistent grammatical style (either all be sentences or all be phrases), but I know what you mean... I'll leave that to you. ;-)

I need to get some sleep now. I'm sure you'll address the existing minor points just fine. Thanks!

**Pull request merged**